kernel/ps2/src/lib.rs

```diff
@@ -1,484 +1,704 @@
-                } else if response == 0xfe {
-                    return Err("Please resend the command to the keyboard");
-                }
-            }
-            Err("command is not acknowledged")
+/// write command to the keyboard and return the result
+fn command_to_keyboard(value: HostToKeyboardCommandOrData) -> Result<(), &'static str> {
+    for _ in 0..3 {
+        let response = send(HostToDevice::Keyboard(value.clone()));
+
+        match response {
+            Ok(Acknowledge) => return Ok(()),
+            Ok(ResendCommand) => continue,
+            _ => continue,
         }
     }
+    Err("keyboard doesn't support the command or there has been a hardware failure")
+}
+
+#[derive(Clone)]
+pub enum HostToMouseCommandOrData {
+    MouseCommand(HostToMouseCommand),
+    SampleRate(SampleRate),
+    MouseResolution(MouseResolution),
+}
+
+// https://wiki.osdev.org/PS/2_Mouse#Mouse_Device_Over_PS.2F2
+#[derive(Clone)]
+pub enum HostToMouseCommand {
+    SetScaling = 0xE6, //1 or 2
+    SetResolution = 0xE8, //[MouseResolution]
+    StatusRequest = 0xE9,
+    SetStreamMode = 0xEA,
+    ReadData = 0xEB,
+    ResetWrapMode = 0xEC,
+    SetWrapMode = 0xEE,
+    SetRemoteMode = 0xF0, //NOTE: same value as SetScancodeSet
+    GetDeviceID = 0xF2, //NOTE: same value as IdentifyKeyboard
+    SampleRate = 0xF3,  //NOTE: same value as SetRepeatRateAndDelay
+    EnableDataReporting = 0xF4, //NOTE: same value as EnableScanning
+    DisableDataReporting = 0xF5, //NOTE: same value as DisableScanning
+    SetDefaults = 0xF6, //NOTE: same
+    ResendByte = 0xFE,  //NOTE: same
+    Reset = 0xFF,       //NOTE: same
 }
 
 /// write command to the mouse and return the result
-pub fn command_to_mouse(value: u8) -> Result<(), &'static str> {
-    // if the mouse doesn't acknowledge the command return error
-    let response = data_to_port2(value);
-    if response != 0xFA {
-        for _x in 0..14 {
-            let response = ps2_read_data();
-
-            if response == 0xFA {
-                //                info!("mouse command is responded !!!!");
-                return Ok(());
-            }
+fn command_to_mouse(value: HostToMouseCommandOrData) -> Result<(), &'static str> {
+    for _ in 0..3 {
+        write_command(WriteByteToPort2InputBuffer);
+        let response = send(HostToDevice::Mouse(value.clone()));
+
+        match response {
+            Ok(Acknowledge) => return Ok(()),
+            _ => continue,
         }
-        warn!("mouse command to second port is not accepted");
-        return Err("mouse command is not responded!!!");
     }
-    //    info!("mouse command is responded !!!!");
-    Ok(())
+    Err("mouse doesn't support the command or there has been a hardware failure")
 }
 
-/// write data to the second ps2 output buffer
-pub fn write_to_second_output_buffer(value: u8) {
-    ps2_write_command(0xD3);
-    unsafe { PS2_PORT.lock().write(value) };
+//TODO: different MousePacketBits for different MouseId?
```

Comment:
yea, i agree it'd be nice to confirm this before merging in this PR. If it's possible, i'd greatly appreciate that.